clang/include/clang/Support/Compiler.h

```diff
@@ -0,0 +1,67 @@
+//===-- clang/Support/Compiler.h - Compiler abstraction support -*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines explicit visibility macros used to export symbols from
+// clang-cpp
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CLANG_SUPPORT_COMPILER_H
+#define CLANG_SUPPORT_COMPILER_H
+
+#include "llvm/Support/Compiler.h"
+
+/// CLANG_ABI is the main export/visibility macro to mark something as
+/// explicitly exported when clang is built as a shared library with everything
+/// else that is unannotated will have internal visibility.
+///
+/// CLANG_EXPORT_TEMPLATE is used on explicit template instantiations in source
+/// files that were declared extern in a header. This macro is only set as a
+/// compiler export attribute on windows, on other platforms it does nothing.
+///
+/// CLANG_TEMPLATE_ABI is for annotating extern template declarations in headers
+/// for both functions and classes. On windows its turned in to dllimport for
+/// library consumers, for other platforms its a default visibility attribute.
+#ifndef CLANG_ABI_GENERATING_ANNOTATIONS
+// Marker to add to classes or functions in public headers that should not have
+// export macros added to them by the clang tool
```

Comment:
```suggestion
// export macros added to them by the clang tool.
```

Reply:
The downside to this is that it's very difficult to know whether or not an API should or shouldn't be exported by the tool. For example, in `Sema`, would we want to export the `ActOn*`  and `Build*` APIs? Probably! But what about the `Check*` APIs? Maybe for most of them. But things like `FillInlineAsmIdentifierInfo`? Probably not.

I seem to recall there being a functional limit to how many interfaces can be exposed in a DLL (I could be remembering wrong though, it's been a while!), so is there harm to over-exposing APIs?

Reply:
To make plugins flexible we probably should export as much as possible. And yes, some of our plugin infrastructure uses even private things from Sema ;)

we could submit patches for them and for some we already have but they cannot be backported across released versions…

Reply:
I think its only really the LLVM DLL built with MSVC or clang-cl without no export inlines that is sitting close to the symbol export limit.

Reply:
Hmmm, are we close enough to the limit we should be worried now, or are we far enough away from the limit that we can worry in the future if/when we hit it? CC @rnk @compnerd @zmodem for extra opinions on whether we need to think about this more before landing

Reply:
The limit is 64K exported symbols for a DLL. I don't know how close Clang would get if we'd try to export everything.

Reply:
I don't have specific insight on the number of clang symbols, but for prior art, LLVM has the [`extract_symbols.py`](https://github.com/llvm/llvm-project/blob/main/llvm/utils/extract_symbols.py) script that tries to export everything. You can see there [was an issue in 2023](https://github.com/llvm/llvm-project/issues/60109) where the symbol limit has been reached.

Generally, DLLs with an explicit export list are usually not at risk of hitting the limit, but it's reasonable to ask "how many exports" does the resulting DLL have and compare to 64K.

Reply:
I managed to get clang-cpp library building without /Zc:dllexportInlines- its at 46748 exported symbols.

Reply:
Would you mind trying an experiment for me? Would you add this to ASTMatchers.h and see how the numbers change for exported symbols:
```
extern const internal::VariadicDynCastAllOfMatcher<Decl, MSPropertyDecl>
    msPropertyDecl;

AST_MATCHER_P(MSPropertyDecl, hasGetter, std::string, Name) {
  return Node.hasGetter() && Node.getGetterId()->isStr(Name);
}
```
I'm asking because AST matchers often have a symbol explosion problem which is easy to overlook. If this adds two exported symbols, I don't worry to much. If it adds 20, it may be more worrying that we're already near 47k out of 65k (not that I think we're going to add that many matchers, just that I worry we'll add more symbols more quickly than we realize).

Reply:
i did have to export some of the AST matcher variables in another PR https://github.com/llvm/llvm-project/pull/110206.

Reply:
Just those lines didn't add any more symbols, I assume you meant add definition of the variable to a source file as well. Doing that only added one extra symbol. I should add there were 200 existing symbols with VariadicDynCastAllOfMatcher in there name that were exported.
[astmatcher symbols.txt](https://github.com/user-attachments/files/17346195/astmatcher.symbols.txt)

Reply:
Thanks! I am surprised those lines didn't add any more symbols -- they're defined inline in the header files. But since that didn't add a pile of new symbols, I think we're probably okay with the amount of headroom we currently have.